src/mesh/replicated_mesh.C

```diff
@@ -1417,5 +1417,151 @@ dof_id_type ReplicatedMesh::n_active_elem () const
                                                  this->active_elements_end()));
 }
 
+std::map<Point, std::vector<std::vector<Point>>>
+ReplicatedMesh::get_boundary_points() const
+{
+  // find number of disconnected subdomains
+  std::vector<Point> interior_points;
+  std::vector<subdomain_id_type> subdomains(max_elem_id() + 1, Elem::invalid_subdomain_id);
+
+  subdomain_id_type subdomain_counter = 0;
+
+  // a stack for visiting elements
+  std::vector<Elem *> list;
+  list.reserve(n_elem());
+
+  dof_id_type visited = 0;
+  do
+  {
+    for (const auto & elem : active_element_ptr_range())
+      if (subdomains[elem->id()] == Elem::invalid_subdomain_id)
+      {
+        list.push_back(elem);
+        interior_points.push_back(elem->centroid());
+        subdomains[elem->id()] = subdomain_counter;
+        break;
+      }
+    while (list.size() > 0)
+    {
+      // pop up an element
+      Elem * elem = list.back(); list.pop_back(); ++visited;
+
+      for (auto s : elem->side_index_range())
+      {
+        Elem * neighbor = elem->neighbor_ptr(s);
+        if (neighbor != nullptr && subdomains[neighbor->id()] == Elem::invalid_subdomain_id)
```

Comment:
This looks like a standard "fill" sweep through a component of the mesh, but note that you can *also* find an unencountered neighbor in the case of AMR meshes, where a coarse active element at a hanging node has an inactive neighbor in that direction.  That could actually be a problem, in fact; a sweep starting from level N elements is never going to find level N+1 or higher elements.  We probably don't care (this is to be run before any adaptive refinement) but it might be a good idea to throw ```libmesh_assert(neighbor->active())``` or something in before putting neighbor on the list in case someone tries to use this method where it's inapplicable.

Reply:
Assertion is a good idea although we possibly can make this algorithm work with adatptively refined meshes.